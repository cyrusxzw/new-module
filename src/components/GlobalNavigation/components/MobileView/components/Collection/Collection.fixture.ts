import type { CollectionProps } from './Collection.types';

const CollectionFixture: CollectionProps = {
  id: 'collection',
  label: 'Collection',
  title: 'Collection',
  items: [
    {
      id: 'ink',
      label: 'Link',
      title: 'Link',
      type: 'link' as const,
      url: '#link',
    },
  ],
};

export { CollectionFixture };
