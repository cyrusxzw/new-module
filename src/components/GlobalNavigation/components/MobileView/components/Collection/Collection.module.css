@import 'styles/index';

.backButton {
  justify-content: flex-start;
}

.directionIcon {
  flex: none;
  margin-right: 12px;

  &.dark {
    polygon {
      fill: var(--color-grey-45);
    }
  }

  &.light {
    polygon {
      fill: var(--color-grey-65);
    }
  }
}

.forward {
  margin-left: 12px;
  transform: rotate(-90deg) translateY(10px);
}

.backwards {
  margin-right: 12px;
  transform: rotate(90deg);
}
