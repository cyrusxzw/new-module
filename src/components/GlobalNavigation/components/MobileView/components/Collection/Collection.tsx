import React from 'react';
import cx from 'classnames';
import { useThemeContext } from '~/contexts';
import { useFocusOnFirst } from '~/customHooks';
import { Transition } from '~/components/Transition';
import { Button } from '~/components/Button';
import { Icon } from '~/components/Icon';
import { useGlobalNavigationStateContext } from '~/components/GlobalNavigation/GlobalNavigation.context';
import { useMobileViewContext } from '~/components/GlobalNavigation/components/MobileView/MobileView.context';
import { ListItem } from '~/components/GlobalNavigation/components/MobileView/components';
import type { CollectionType } from './Collection.types';
import compositionStyles from '../../MobileView.module.css';
import styles from './Collection.module.css';

const Collection: CollectionType = ({
  backLabel,
  id,
  isVisible,
  items,
  label,
  promotion,
  shouldHideTopCollection = false,
  title,
}) => {
  const {
    activeCollectionId,
    setActiveCollectionId,
  } = useGlobalNavigationStateContext();
  const { onBackButtonClick } = useMobileViewContext();
  const currentTheme = useThemeContext(null, 'dark');

  const isActive = activeCollectionId === id;
  const [listRef] = useFocusOnFirst(
    isActive,
    'a:not([tabindex="-1"]), button:not([tabindex="-1"])',
  );
  const handleOnClick = () => setActiveCollectionId(id);

  const handleOnBackButtonClick = () => {
    setActiveCollectionId('top');
    onBackButtonClick();
  };

  const listClassSet = cx(
    compositionStyles.list,
    compositionStyles.sublist,
    compositionStyles.slide,
  );

  const forwardIconClassSet = cx(
    styles.directionIcon,
    styles.forward,
    styles[currentTheme],
  );

  const backButtonClassSet = cx(
    compositionStyles.itemElement,
    styles.backButton,
  );

  const backIconClassSet = cx(
    styles.directionIcon,
    styles.backwards,
    styles[currentTheme],
  );

  return (
    <>
      {!shouldHideTopCollection && (
        <Button
          aria={{
            expanded: isActive,
            haspopup: true,
            hidden: !isActive,
          }}
          className={cx(compositionStyles.itemElement)}
          isInline={true}
          onClick={handleOnClick}
          tabIndex={activeCollectionId === 'top' && isVisible ? null : -1}
          theme={currentTheme}
          title={title}
        >
          {label}
          <Icon
            className={forwardIconClassSet}
            height={12}
            name="chevron"
            theme={currentTheme}
            width={12}
          />
        </Button>
      )}

      <Transition
        isActive={isActive && !!items && !!items.length}
        type="slideLeft"
      >
        <ul
          aria-hidden={!isActive}
          aria-label="submenu"
          className={listClassSet}
          ref={listRef}
        >
          <li className={compositionStyles.listItem}>
            <Button
              aria={{ hidden: !isActive, label: 'back' }}
              className={backButtonClassSet}
              isInline={true}
              onClick={handleOnBackButtonClick}
              tabIndex={isActive ? null : -1}
              theme={currentTheme}
              title={backLabel}
            >
              <Icon
                className={backIconClassSet}
                height={14}
                name="chevron"
                width={14}
              />{' '}
              {backLabel || label}
            </Button>
          </li>

          {items.map((props) => (
            <ListItem
              isVisible={isActive}
              itemProps={props}
              key={props.label}
            />
          ))}

          {promotion && <ListItem isVisible={isActive} itemProps={promotion} />}
        </ul>
      </Transition>
    </>
  );
};

export { Collection };
