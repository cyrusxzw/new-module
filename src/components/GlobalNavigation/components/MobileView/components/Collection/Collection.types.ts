import type { ComponentWithoutChildren } from '~/types';
import type {
  CategoryItemTrackingWithMobileAction,
  Clickable,
  Link,
  NestedCollection,
  NotableNestedCollection,
  MenuItemTrackingWithMobileAction,
  Promotion,
} from '~/components/GlobalNavigation/GlobalNavigation.types';

type CollectionProps = Clickable & {
  backLabel?: string;
  isVisible?: boolean;
  items: (NestedCollection | Link | NotableNestedCollection)[];
  promotion?: Promotion;
  shouldHideTopCollection?: boolean;
};

type CollectionType = ComponentWithoutChildren<CollectionProps>;

type CategoryItemTrackingWithMobileActionType = CategoryItemTrackingWithMobileAction;

type MenuItemTrackingWithMobileActionType = MenuItemTrackingWithMobileAction;

export type {
  CategoryItemTrackingWithMobileActionType,
  CollectionType,
  CollectionProps,
  MenuItemTrackingWithMobileActionType,
};
