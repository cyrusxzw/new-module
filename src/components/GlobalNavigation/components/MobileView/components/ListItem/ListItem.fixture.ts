import type { ListItemProps } from './ListItem.types';

/* TODO{issue-10}: Improve fixture to be more reflective of tracking data
 * Currently implemented type to satisfy typescript for this fixture
 */

type NavTrackingPropsType = {
  menuType: 'Shop' | 'Read';
  panel: 'Panel 1' | 'Panel 2';
};

const navTrackingProps: NavTrackingPropsType = {
  menuType: 'Shop',
  panel: 'Panel 1',
};

const ListItemFixture: ListItemProps = {
  itemProps: {
    id: 'nested-collection',
    label: 'Nested Collection',
    title: 'Nested Collection',
    type: 'nested-collection' as const,
    items: [
      {
        id: 'nested-link',
        label: 'Nested Link',
        title: 'Nested Link',
        type: 'link' as const,
        url: '#nested-link',
        ...navTrackingProps,
      },
    ],
  },
  isVisible: true,
  isNestedItem: false,
};

export { ListItemFixture };
