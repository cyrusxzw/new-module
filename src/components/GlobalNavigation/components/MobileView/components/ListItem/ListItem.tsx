import React from 'react';
import cx from 'classnames';
import {
  Collection,
  Link,
  PromotionCard,
  NestedCollection,
} from '../../components';
import type { ListItemType } from './ListItem.types';
import compositionStyles from '../../MobileView.module.css';

const ListItem: ListItemType = ({
  isNestedItem,
  isTopItem = false,
  isVisible = false,
  itemProps,
}) => {
  let returnElement = null;

  if (itemProps.type === 'collection') {
    returnElement = <Collection {...{ ...itemProps, isVisible }} />;
  } else if (itemProps.type === 'nested-collection') {
    returnElement = <NestedCollection {...{ ...itemProps, isVisible }} />;
  } else if (itemProps.type === 'promotion-card') {
    returnElement = <PromotionCard {...{ ...itemProps, isVisible }} />;
  } else if (itemProps.type === 'link') {
    returnElement = (
      <Link {...{ ...itemProps, isVisible, isNestedItem, isTopItem }} />
    );
  } else {
    return null;
  }

  const classSet = cx(compositionStyles.listItem, {
    [compositionStyles.nestedItem]: itemProps.type === 'nested-collection',
    [compositionStyles.nestedCollectionLink]:
      itemProps.type === 'link' && isNestedItem,
  });

  return <li className={classSet}>{returnElement}</li>;
};

export { ListItem };
