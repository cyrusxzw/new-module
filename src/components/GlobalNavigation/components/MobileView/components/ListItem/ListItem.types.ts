import type { ComponentWithChildren } from '~/types';
import type {
  Promotion,
  Collection,
  Link,
  Read,
  NestedCollection,
  NotableNestedCollection,
} from '~/components/GlobalNavigation/GlobalNavigation.types';

type ListItemProps = {
  isNestedItem?: boolean;
  isTopItem?: boolean;
  isVisible?: boolean;
  itemProps:
    | NestedCollection
    | Collection
    | Link
    | Promotion
    | NotableNestedCollection
    | Read;
};

type ListItemType = ComponentWithChildren<ListItemProps>;

export type { ListItemType };
