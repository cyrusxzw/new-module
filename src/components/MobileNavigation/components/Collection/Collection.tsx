import React from 'react';
import cx from 'classnames';
import { useThemeContext } from '~/contexts';
import { useFocusOnFirst } from '~/customHooks';
import { Transition } from '~/components/Transition';
import { Button } from '~/components/Button';
import { Icon } from '~/components/Icon';
import { useMobileNavigationContext } from '~/components/MobileNavigation/MobileNavigation.context';
import { ListItem } from '~/components/MobileNavigation/components';
import type { CollectionType } from './Collection.types';
import compositionStyles from '~/components/MobileNavigation/MobileNavigation.module.css';
import styles from './Collection.module.css';

const Collection: CollectionType = ({
  id,
  isActive: isVisible,
  items,
  title,
  label,
}) => {
  const currentTheme = useThemeContext(null, 'dark');
  const {
    activeCollectionId,
    onCollectionClick,
    onBackButtonClick,
  } = useMobileNavigationContext();
  const isActive = activeCollectionId === id;
  const [listRef] = useFocusOnFirst(
    isActive,
    'a:not([tabindex="-1"]), button:not([tabindex="-1"])',
  );
  const handleOnClick = () => onCollectionClick(id);

  const listClassSet = cx(
    compositionStyles.list,
    compositionStyles.sublist,
    styles.slide,
  );

  const forwardIconClassSet = cx(
    styles.directionIcon,
    styles.forward,
    styles[currentTheme],
  );

  const backButtonClassSet = cx(
    compositionStyles.itemElement,
    compositionStyles.ornamentalWrapper,
    styles.backButton,
  );

  const backIconClassSet = cx(
    styles.directionIcon,
    styles.backwards,
    styles[currentTheme],
  );

  return (
    <>
      <Button
        aria={{
          expanded: isActive,
          haspopup: true,
          hidden: !isActive,
        }}
        aria-expanded={isActive}
        aria-haspopup="true"
        className={cx(
          compositionStyles.itemElement,
          compositionStyles.ornamentalWrapper,
        )}
        isInline={true}
        onClick={handleOnClick}
        tabIndex={activeCollectionId === 'top' && isVisible ? null : -1}
        theme={currentTheme}
        title={title}
      >
        <span className={compositionStyles.ornamentalHover}>{label}</span>{' '}
        <Icon
          className={forwardIconClassSet}
          height={12}
          name="chevron"
          theme={currentTheme}
          width={12}
        />
      </Button>

      <Transition
        isActive={isActive && !!items && !!items.length}
        type="slideLeft"
      >
        <ul
          aria-hidden={!isActive}
          aria-label="submenu"
          className={listClassSet}
          ref={listRef}
        >
          <li className={compositionStyles.listItem}>
            <Button
              aria={{ hidden: !isActive, label: 'back' }}
              className={backButtonClassSet}
              isInline={true}
              onClick={onBackButtonClick}
              tabIndex={isActive ? null : -1}
              theme={currentTheme}
              title={title}
            >
              <Icon
                className={backIconClassSet}
                height={14}
                name="chevron"
                width={14}
              />{' '}
              <span className={compositionStyles.ornamentalHover}>
                All Collections
              </span>
            </Button>
          </li>

          {items.map((props) => (
            <ListItem isActive={isActive} itemProps={props} key={props.label} />
          ))}
        </ul>
      </Transition>
    </>
  );
};

export { Collection };
