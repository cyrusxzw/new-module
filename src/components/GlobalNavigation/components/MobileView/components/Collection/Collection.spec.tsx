/* eslint-disable init-declarations */
import React from 'react';
import { render, screen } from '@testing-library/react';
import { Collection } from './Collection';
import { CollectionFixture } from './Collection.fixture';
import {
  GlobalNavigationStateContextProvider,
  useGlobalNavigationStateContext,
} from '~/components/GlobalNavigation/GlobalNavigation.context';
import {
  MobileViewContextProvider,
  useMobileViewContext,
} from '~/components/GlobalNavigation/components/MobileView/MobileView.context';

import * as ListItemFile from '../ListItem/ListItem';

jest.mock('~/components/GlobalNavigation/GlobalNavigation.context');
jest.mock(
  '~/components/GlobalNavigation/components/MobileView/MobileView.context',
);

describe('<GlobalNavigation.MobileView.Collection />', () => {
  let listItemSpy: jest.SpyInstance;

  beforeEach(() => {
    jest.clearAllMocks();
    listItemSpy = jest
      .spyOn(ListItemFile, 'ListItem')
      .mockImplementation(() => null);
    (useGlobalNavigationStateContext as jest.Mock).mockReturnValue({
      activeCollectionId: CollectionFixture.id,
      setActiveCollectionId: jest.fn(),
    });
    (useMobileViewContext as jest.Mock).mockReturnValue({
      setActiveNestedCollectionIds: jest.fn(),
    });
  });

  const TestBed = (props) => {
    return (
      <GlobalNavigationStateContextProvider>
        <MobileViewContextProvider>
          <Collection {...CollectionFixture} {...props} />
        </MobileViewContextProvider>
      </GlobalNavigationStateContextProvider>
    );
  };

  it('should render base component correctly', () => {
    const { container } = render(<TestBed />);

    expect(container).not.toBeEmptyDOMElement();
  });

  it('should not render top collection when hidden', () => {
    render(<TestBed shouldHideTopCollection={true} />);
    const buttons = screen.getAllByRole('button');

    expect(buttons.length).toEqual(1);
    expect(buttons[0]).not.toHaveAttribute('title', CollectionFixture.title);
  });

  it('should change active collection id on clicking top level collection', () => {
    (useGlobalNavigationStateContext as jest.Mock).mockReturnValue({
      activeCollectionId: 'top',
      setActiveCollectionId: jest.fn(),
    });

    const {
      activeCollectionId,
      setActiveCollectionId,
    } = useGlobalNavigationStateContext();

    render(<TestBed id={activeCollectionId} />);

    const topCollectionButton = screen.getByTitle(CollectionFixture.title);
    topCollectionButton.click();

    expect(setActiveCollectionId).toHaveBeenCalledTimes(1);
  });

  it('should have tabbable top collection button if visible', () => {
    (useGlobalNavigationStateContext as jest.Mock).mockReturnValue({
      activeCollectionId: 'top',
      setActiveCollectionId: jest.fn(),
    });

    const { activeCollectionId } = useGlobalNavigationStateContext();
    render(<TestBed id={activeCollectionId} isVisible={true} />);

    const topCollectionButton = screen.getByTitle(CollectionFixture.title);

    expect(topCollectionButton).not.toHaveAttribute('tabIndex');
  });

  it('should not have tabbable collection back button if not active', () => {
    (useGlobalNavigationStateContext as jest.Mock).mockReturnValue({
      activeCollectionId: 'top',
      setActiveCollectionId: jest.fn(),
    });

    render(<TestBed />);

    const nestedCollectionButton = screen.getAllByRole('button', {
      hidden: true,
    })[1];

    expect(nestedCollectionButton).toHaveAttribute('tabIndex', '-1');
  });

  it('should return to top level collection by clicking back button', () => {
    render(<TestBed shouldHideTopCollection={true} />);

    const { setActiveNestedCollectionIds } = useMobileViewContext();

    const nestedCollectionButton = screen.getByRole('button');
    nestedCollectionButton.click();

    expect(setActiveNestedCollectionIds).toHaveBeenCalledTimes(1);
    expect(setActiveNestedCollectionIds).toHaveBeenCalledWith([]);
  });

  it('should render promotion list item', () => {
    render(<TestBed promotion={CollectionFixture.promotion} />);

    // Always either 1 or 0
    const promotionCount = CollectionFixture.promotion ? 1 : 0;

    const listItemCount = CollectionFixture.items.length + promotionCount;

    expect(listItemSpy).toHaveBeenCalledTimes(listItemCount);
  });
});
