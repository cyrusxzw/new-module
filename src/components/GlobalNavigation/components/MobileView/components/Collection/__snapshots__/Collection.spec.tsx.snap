// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<GlobalNavigation.MobileView.Collection /> should render base component correctly 1`] = `
<div>
  <button
    aria-expanded="true"
    aria-haspopup="true"
    aria-hidden="false"
    class="base dark itemElement"
    tabindex="-1"
    title="Collection"
    type="button"
  >
    Collection
    <svg
      aria-labelledby="chevron-1"
      class="base chevron dark directionIcon forward dark"
      data-testid="data-testid-Icon"
      focusable="false"
      height="12"
      role="img"
      style="height: 12px; width: 12px;"
      tabindex="-1"
      viewBox="0 0 50 50"
      width="12"
    >
      <title
        id="chevron-1"
      >
        chevron
      </title>
      <g>
        <polygon
          class="chevron"
          points="25,31.3 4.2,10.5 0.1,14.6 25,39.5 50,14.6 45.9,10.5"
        />
      </g>
    </svg>
  </button>
  <ul
    aria-hidden="false"
    aria-label="submenu"
    class="list sublist slide"
  >
    <li
      class="listItem"
    >
      <button
        aria-hidden="false"
        aria-label="back"
        class="base dark itemElement backButton"
        title="Return from collection"
        type="button"
      >
        <svg
          aria-labelledby="chevron-2"
          class="base chevron dark directionIcon backwards dark"
          data-testid="data-testid-Icon"
          focusable="false"
          height="14"
          role="img"
          style="height: 14px; width: 14px;"
          tabindex="-1"
          viewBox="0 0 50 50"
          width="14"
        >
          <title
            id="chevron-2"
          >
            chevron
          </title>
          <g>
            <polygon
              class="chevron"
              points="25,31.3 4.2,10.5 0.1,14.6 25,39.5 50,14.6 45.9,10.5"
            />
          </g>
        </svg>
         
        Return from collection
      </button>
    </li>
  </ul>
</div>
`;
